### ISSUE TITLE
Data splits on the training scripts

### ISSUE THREAD

**Issue by RongKaiWeskerMA** [TeaPearce/Counter-Strike_Behavioural_Cloning#26]
Dear Tim,

Thanks for contributing the community! I have some confusion about the data splits strategy in the [dm_train_model.py](https://github.com/TeaPearce/Counter-Strike_Behavioural_Cloning/blob/main/dm_train_model.py)

`data_list1 = [str(x1)+'-'+str(x2) for x1 in np.arange(starting_num,highest_num+1) for x2 in np.arange(0,1000-N_TIMESTEPS-int(N_JITTER),N_TIMESTEPS)]
data_list2 = [str(x1)+'-'+str(x2) for x1 in np.arange(starting_num,highest_num+1) for x2 in np.arange(0,1000-N_TIMESTEPS-int(N_JITTER),N_TIMESTEPS)]
data_list3 = [str(x1)+'-'+str(x2) for x1 in np.arange(starting_num,highest_num+1) for x2 in np.arange(0,1000-N_TIMESTEPS-int(N_JITTER),N_TIMESTEPS)]
data_list4 = [str(x1)+'-'+str(x2) for x1 in np.arange(starting_num,highest_num+1) for x2 in np.arange(0,1000-N_TIMESTEPS-int(N_JITTER),N_TIMESTEPS)]`

`data_list1 = fn_subselect(data_list1, subselect_helper_dict, SUB_PROB, True)
    data_list2 = fn_subselect(data_list2, subselect_helper_dict, SUB_PROB, True)
    data_list3 = fn_subselect(data_list3, subselect_helper_dict, SUB_PROB, True)
    data_list4 = fn_subselect(data_list4, subselect_helper_dict, SUB_PROB, True)`

1. the data_list1 ~ data_list4 are exactly the same, but the fn_subselect filters out some unuseful segments on all 4 data_lists. However, I reckon, the samples after subselection still overlap heavily. Could you please explain a bit on why we are doing this? I think we can just filter once.

2. Morevoer, the training code on the pretrain data is sequentially train the model on these 4 splits:
`hist = model.fit(training_generator1,epochs=1,workers=4,verbose=1,use_multiprocessing=True, max_queue_size=20)
        tp_save_model(model, save_dir, model_name+iter_letter+'1')
        hist = model.fit(training_generator2,epochs=1,workers=4,verbose=1,use_multiprocessing=True, max_queue_size=20)
        tp_save_model(model, save_dir, model_name+iter_letter+'2')
        hist = model.fit(training_generator3,epochs=1,workers=4,verbose=1,use_multiprocessing=True, max_queue_size=20)
        tp_save_model(model, save_dir, model_name+iter_letter+'3')
        hist = model.fit(training_generator4,epochs=1,workers=4,verbose=1,use_multiprocessing=True, max_queue_size=20)
        tp_save_model(model, save_dir, model_name+iter_letter+'4')`

I reckon if these splits are distinct from each other, we can just literally concatenate the 4 lists and train the model once?

**Comment by TeaPearce** [comment]
my memory is properly hazy on this. I _think_ the only reason I did it is that calling .fit method in keras committed to training the model for one whole epoch, and I didn't want to wait that long in between model saves, so created four dataloaders of 1/4 the size.

if you're training models in 2024 anyway you probably want to use torch

**Comment by RongKaiWeskerMA** [comment]
One more question. I realize that after subselection data_list1 - data_list4 still overlap 90 % between each pair of them. If you sequentially train the model on them I would reckon its almost the same as training the model on the same dataset for 4 epochs with shuffle? 

even though sequential training would result in forgetting, the 4 subsets are almost identical.

**Comment by tuan124816** [comment]
Hi @RongKaiWeskerMA ,

I noticed from your comment that you were able to run this repository on the current Counter-Strike 2. I’m also trying to get it to work but have struggled to adapt.

Would you mind sharing if you had to make significant changes to the source code to get it running on CS2? If possible, could you share any details or modifications you made? It would be a huge help!

Also, if it’s okay with you, could I have your email to reach out for further questions?

Thanks a lot for your help in advance!

**Comment by RongKaiWeskerMA** [comment]
> Hi @RongKaiWeskerMA ,
> 
> I noticed from your comment that you were able to run this repository on the current Counter-Strike 2. I’m also trying to get it to work but have struggled to adapt.
> 
> Would you mind sharing if you had to make significant changes to the source code to get it running on CS2? If possible, could you share any details or modifications you made? It would be a huge help!
> 
> Also, if it’s okay with you, could I have your email to reach out for further questions?
> 
> Thanks a lot for your help in advance!

Hi @tuan124816, you need to modify the window capture scripts to be able to run this repo with cs2, since I remember the old capture code didn't work with cs2. However, the csgo agent didn't navigate nor shooting in the cs2. I guess it's due to the update of graphics (illumination, lighting, and scene details), which affect a lot. The best way I guess to make it run with cs2 is to record cs2 data and fine-tune.

**Comment by tuan124816** [comment]
> > Hi @RongKaiWeskerMA ,
> > I noticed from your comment that you were able to run this repository on the current Counter-Strike 2. I’m also trying to get it to work but have struggled to adapt.
> > Would you mind sharing if you had to make significant changes to the source code to get it running on CS2? If possible, could you share any details or modifications you made? It would be a huge help!
> > Also, if it’s okay with you, could I have your email to reach out for further questions?
> > Thanks a lot for your help in advance!
> 
> Hi @tuan124816, you need to modify the window capture scripts to be able to run this repo with cs2, since I remember the old capture code didn't work with cs2. However, the csgo agent didn't navigate nor shooting in the cs2. I guess it's due to the update of graphics (illumination, lighting, and scene details), which affect a lot. The best way I guess to make it run with cs2 is to record cs2 data and fine-tune.

Hi @RongKaiWeskerMA,

Thanks so much for your response and suggestions! I’ve been trying to adapt this repository to CS2 as well, but I’ve run into some challenges.

I’ve already created a new screen capture script and adjusted my CS2 GSI config file, but it executes very slowly when I run the code. Could you kindly share your updated capture screen scripts and CS2 GSI config file? It would be incredibly helpful to see how you’ve optimized it.

Also, are you using the same environment setup, or did you make any additional changes to improve compatibility or performance?

If you'd like to see my capture screen code and GSI config file, feel free to let me know, and I’d be happy to share them here in this comment section.

Looking forward to your insights. Your help would mean a lot!

**Comment by ysb003** [comment]
@tuan124816 It seems that the screenshot process in this instance might slow down the speed. My approach is to capture a portion of the entire desktop and then transfer it to the memory for rendering, only rendering the parts that have been cropped in memory. This way, the speed should increase. I am very keen to discuss with you that my CS2 GSI configuration is almost identical to CS1, as I have only updated a part of the memory and it is not fully operational yet. I am not sure if it is correct, and I hope to receive your guidance to see if there are any misconfigurations.

However, during the reasoning process, everything seems to be problem-free. I hope to obtain your email address for this challenging task, as I am willing to invest time to overcome it.

**Comment by RongKaiWeskerMA** [comment]
Hi @tuan124816 , I'm not following the same gsi configuration as author provided, since I can record the mouse information. The only thing I need for the GSI is to obtain the reward for the model to train (last output from the model).

**Comment by ysb003** [comment]
@RongKaiWeskerMA Dear Sir,

I am currently stuck in the training process because CSGO2 also resets the mouse position information. My approach is to follow the original author's idea to obtain it through memory. I am using a method to read memory by honestly calculating the offset with tools, which I am still learning and have not yet completed. How do you record mouse movements? Do you need to convert it into a specific format after obtaining it? Does obtaining the mouse movement mean that I can stop using memory to get information and proceed directly to the next step of training?

**Comment by RongKaiWeskerMA** [comment]
Hi @ysb003 Try the pynput package (on_move) function specifically. See if this helps

**Comment by ysb003** [comment]
非常感谢，我会尝试使用(on_move) 直接推理出鼠标的位移情况并更新到训练集里面，如果训练完成我将会在这里汇报一下具体情况

**Comment by ysb003** [comment]
@RongKaiWeskerMA I will try to record and update it into the training set. If successful, I will reply here. Thank you for your help.

**Comment by RongKaiWeskerMA** [comment]
puzzle solved, tensorflow/keras tends to load the dataset into gpu at once. The author trained the model with v100 which does not have enough memory to do so. 


Hopefully my understanding is correctly.
